### PR TITLE
EZP-24253: Implement framework for ContentType creation

### PR DESCRIPTION
**Pull request by lolautruche**
> https://jira.ez.no/browse/EZP-24253

Modifications to allow flexible implementation of ContentType forms.

* [x] Allow ContentType creation without FieldDefinition
* [x] Fix TextLine validator default values
* [x] Fix tests
* [x] Add BC note

### CODE REVIEW

**Comment by lolautruche** [comment]
Review ping @andrerom @bdunogier @pspanja

**Comment by bdunogier** [comment]
+1

File: eZ/Publish/Core/FieldType/TextLine/Type.php [hunk 1]
```diff
@@ -31,7 +31,7 @@ class Type extends FieldType
             ),
             'maxStringLength' => array(
                 'type' => 'int',
-                'default' => false
+                'default' => null
```
Comment (pspanja) [inline eZ/Publish/Core/FieldType/TextLine/Type.php]: Does this affect REST? Ping @dpobel.

File: eZ/Publish/Core/Repository/ContentTypeService.php [hunk 2]
```diff
@@ -552,14 +552,6 @@ protected function validateInputContentTypeCreateStruct( APIContentTypeCreateStr
             );
         }
 
-        if ( empty( $contentTypeCreateStruct->fieldDefinitions ) )
```
Comment (pspanja) [inline eZ/Publish/Core/Repository/ContentTypeService.php]: This removal goes against what is documented on the interface, so it is really an API change. So this needs updated interface docblocks and BC notes.

File: eZ/Publish/Core/FieldType/TextLine/Type.php (hunk 1 above)
Comment (lolautruche) [inline eZ/Publish/Core/FieldType/TextLine/Type.php]: Don't think the change will break anything. However, `validateValidatorConfiguration()` may report new errors if one is passing `false`.

File: eZ/Publish/Core/Repository/ContentTypeService.php (hunk 2 above)
Comment (lolautruche) [inline eZ/Publish/Core/Repository/ContentTypeService.php]: Yep, you're right. Forgot that.

Comment (lolautruche) [inline eZ/Publish/Core/Repository/ContentTypeService.php]: Fixed

**Comment by lolautruche** [comment]
BC notes added and API change reflected in PHPDoc

File: eZ/Publish/Core/FieldType/TextLine/Type.php (hunk 1 above)
Comment (pspanja) [inline eZ/Publish/Core/FieldType/TextLine/Type.php]: I took a quick look at `PlatformUIBundle`, it seems it should be ok.

**Comment by pspanja** [comment]
Looks good, +1.

**Comment by andrerom** [comment]
+1

*Assuming* the following, so please please verify ( @lolautruche / @pspanja ):
- that we indeed validate that content type draft has a field before you are allowed to publish (interface says so already, did not check impl)
-- because at a minimum there should be one field[Def] that maps to nameSchema

**Comment by pspanja** [comment]
@andrerom Yes, that is implemented: https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Repository/ContentTypeService.php#L1966

**Comment by lolautruche** [comment]
@andrerom Already checked that and it's validated during ContentType publish.

**Comment by andrerom** [comment]
Ok cool, @pspanja where do we validate nameschema/urlschema? it's currently done in persistence right? so meaning it is hopefully done on publish as well.

**Comment by lolautruche** [comment]
@andrerom Even it's worth checking, this is unrelated anyway :wink:

**Comment by pspanja** [comment]
@andrerom 

It is created on publish from the first field definition if empty, otherwise it is not validated.
But as @lolautruche said, that is separate if we want to do it.

**Comment by lolautruche** [comment]
> It is created on publish from the first field definition if empty, otherwise it is not validated.

I guess what @andrerom is asking for, is when you're adding a custom schema.

**Comment by andrerom** [comment]
yes, it is unrelated, just wanted to bring the big picture ;)

**Comment by pspanja** [comment]
@lolautruche 

It is not validated :) The way the thing works, there is no invalid schema. It tries to resolves recognized tokens in the schema string, tolerates errors with using empty strings, and simply returns the result.

Publishing the alias that resolves into the empty string uses default alias string (node_$locationId), and for content name empty string is accepted.

**Comment by lolautruche** [comment]
OK.

Waiting for Travis to give its final +1 before merging :-)

**Pull request merged**